hearth/lib/hearth/http2/connection.rb

```diff
@@ -0,0 +1,259 @@
+      def initialize(options = {})
+        OPTIONS.each_pair do |opt_name, default_value|
+          value = options[opt_name].nil? ? default_value : options[opt_name]
+          instance_variable_set("@#{opt_name}", value)
+        end
+
+        @h2_client = ::HTTP2::Client.new
+        @mutex = Mutex.new
+
+        @socket = create_tcp_connection(options)
+        register_h2_callbacks
+        @state = :connected
+        @healthy = true
+
+        @streams = {} # map of stream.id -> stream
+        @errors = []
+        start_connection_thread
+      end
+
+      OPTIONS.each_key do |attr_name|
+        attr_reader(attr_name)
+      end
+
+      alias verify_peer? verify_peer
+
+      def stale?
+        !@healthy || @state != :connected
+      end
+
+      # return a new stream, or nil when max streams is exceeded
+      def new_stream
+        return unless @streams.size < @max_concurrent_streams
+
+        begin
+          stream = @h2_client.new_stream
+          @streams[stream.id] = stream
+          stream
+        rescue ::HTTP2::Error::StreamLimitExceeded
+          nil # exceeded our max streams from remote (server) settings
+        end
+      end
+
+      def close_stream(stream)
+        @mutex.synchronize do
+          @streams.delete(stream.id)
+          stream.close
+        end
+      end
+
+      # all underlying streams will be closed
+      def close
+        @mutex.synchronize do
+          log_debug('closing connection')
+          @status = :closed
+          @healthy = false
+
+          @streams.each_value(&:close)
+          @streams = {}
+          @thread.kill
+
+          @socket&.close
+        end
+      end
+      alias finish close
+
+      private
+
+      # rubocop:disable Metrics
+      def start_connection_thread
+        @mutex.synchronize do
+          @thread = Thread.new do
+            while !@socket.closed? && !@socket.eof?
+              begin
+                data = @socket.read_nonblock(CHUNKSIZE)
+                @h2_client << data
+              rescue IO::WaitReadable
+                begin
+                  if @socket.wait_readable(read_timeout)
+                    # available, retry to start reading
+                    retry
+                  else
+                    log_debug('socket connection read time out')
+                    close
+                  end
+                rescue StandardError
+                  # error can happen when closing the socket
+                  # while it's waiting for read
+                  close
+                end
+              rescue StandardError => e
+                @errors << e
+                close
+                raise e
+              end
+            end
+            close
+          end
+          @thread.abort_on_exception = true
+        end
+      end
+      # rubocop:enable Metrics
+
+      def create_tcp_connection(options)
+        endpoint = options[:endpoint]
+        tcp, addr = tcp_socket(endpoint)
```

Comment:
Is there no way to connect to http2 just using a host name (have it do resolution itself?)

Reply:
No - we need to create an manage the socket ourselves.  If a resolver is not configured we are able to use the `:Socket.getaddrinfo` but we still need to call that ourselves.